### PR TITLE
feat(stenciltest): introduce testing framework

### PR DESCRIPTION
**Pull request by jaredallard**
## What this PR does / why we need it

This PR introduces a testing framework for stencil powered by `go test`. This is done by providing a public package called `stenciltest`. This is meant to be called by a test in a stencil template repository. Take, for example, testing the `README.md.tpl` template in [stencil-base](https://github.com/getoutreach/stencil-base/blob/main/README.md.tpl). This would be done by first creating a light `go.mod` in the repository like so:

```go
module github.com/getoutreach/stencil-base

require github.com/getoutreach/stencil v0.0.0
```

At this point we can create a test file anywhere, for example we can create `main_test.go`. Here we'll use the `stenciltest` package to test the `README.md` template when adding the `oss` argument.

```go
package main

import (
	"testing"

	"github.com/getoutreach/stencil/pkg/stenciltest"
)

func TestReadmeOSSRender(t *testing.T) {
	st := stenciltest.New(t, "README.md.tpl", "helpers.tpl")
	st.Args(map[string]interface{}{ 
		"oss": true,
	})
	st.Run(false)
}
```

Lets run the test::

```shell
日本 stencil-base ❯ gotestsum -f dots-v2 ./...

  30ms . ✖✖✖

 3 tests, 3 failures in 1.082s

=== Failed
=== FAIL: . TestReadmeRender/README.md.tpl/README.md (0.00s)
    stenciltest.go:105: snapshot created for test TestReadmeRender-README.md.tpl-README.md
        --- FAIL: TestReadmeRender/README.md.tpl/README.md (0.00s)

=== FAIL: . TestReadmeRender/README.md.tpl (0.01s)
    stenciltest.go:108: Generated file "README.md" did not match snapshot
    --- FAIL: TestReadmeRender/README.md.tpl (0.01s)

=== FAIL: . TestReadmeRender (0.02s)

DONE 3 tests, 3 failures in 1.124s
```

When this test is ran the first time, it'll error because a snapshot was not found in the `.snapshot` subdirectory of the current test. However, it created a snapshot. On the next run it'll succeed because it matches. Nice!

```shell
日本 stencil-base ❯ gotestsum -f dots-v2 ./...                                                                                             main


  40ms . ···

 3 tests in 1.420s

DONE 3 tests in 1.458s
```

If we change `oss` to true, it'll fail:

```shell
日本 stencil-base ❯ gotestsum -f dots-v2 ./...                                                                                             main


  40ms . ✖✖✖

 3 tests, 3 failures in 2.130s

=== Failed
=== FAIL: . TestReadmeRender/README.md.tpl/README.md (0.00s)
    stenciltest.go:105: snapshot not equal:
        --- Previous
        +++ Current
        @@ -2,3 +2,3 @@
         # testing
        -[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white)](https://pkg.go.dev/github.com/getoutreach/testing)
        +[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white)](https://engdocs.outreach.cloud/github.com/getoutreach/testing)
         [![Generated via Bootstrap](https://img.shields.io/badge/Outreach-Bootstrap-%235951ff)](https://github.com/getoutreach/bootstrap)

        --- FAIL: TestReadmeRender/README.md.tpl/README.md (0.00s)

=== FAIL: . TestReadmeRender/README.md.tpl (0.02s)
    stenciltest.go:108: Generated file "README.md" did not match snapshot
    --- FAIL: TestReadmeRender/README.md.tpl (0.02s)

=== FAIL: . TestReadmeRender (0.02s)

DONE 3 tests, 3 failures in 2.171s
```

That's basically it! The snapshots can be viewed in the `.snapshots` folder, and match the format of `TestName-Template-File`. Currently, you can only test one template per go test. So, multiple argument tests need multiple tests or use sub-tests for table tests.

snapshot

## Jira ID

[XX-XX]

### CODE REVIEW

File: pkg/stenciltest/stenciltest.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+type Template struct {
+	// path is the path to the template.
+	path string
+
+	// aditionalTemplates is a list of additional templates to add to the renderer,
+	// but not to snapshot.
+	additionalTemplates []string
+
+	// manifestPath is the path to the manifest file.
+	// IDEA(jaredallard): We should try to find this automatically
```
Comment (jaredallard) [inline pkg/stenciltest/stenciltest.go:36]:
```suggestion
	// m is the template repository manifest for this test
```

File: pkg/stenciltest/stenciltest.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+// func TestCanRender(t *testing.T) {
+// 	st := stenciltest.New("README.md.tpl")
+// 	st.Args(map[string]interface{
+// 		"name": "do a thing",
+// 	})
+
+// 	st.Run(t)
+// }
```
Comment (jaredallard) [inline pkg/stenciltest/stenciltest.go:124]:
```suggestion
```

I did a great job cleaning up this code I swear

File: pkg/stenciltest/stenciltest.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		if err != nil {
+			got.Errorf("failed to create module from template %q", t.path)
+		}
+
+		mf := &configuration.ServiceManifest{Name: "testing", Arguments: t.args,
+			Modules: []*configuration.TemplateRepository{{Name: m.Name}}}
+		st := codegen.NewStencil(mf, []*modules.Module{m})
+
+		tpls, err := st.Render(context.Background(), logrus.New())
+		if err != nil {
```
Comment (ChrisPRobinson) [inline pkg/stenciltest/stenciltest.go:91]: Will there be a case where the template owner will want to validate that an error is received?

For Temporal as an example we previously had as part of the template validation of the versions and failed if it was an unsupported one. 

This is probably an edge-case the scenario I'm pointing out but in a roundabout way I'm wondering if we'd want to validate an error occurs in certain templates.

Comment (jaredallard) [reply pkg/stenciltest/stenciltest.go:91]: That's a fair ask! That makes sense, I'll add that in.

**Review by ChrisPRobinson — APPROVED** [review]
LGTM

**Review by bhogatemadhura — APPROVED** [review]

File: pkg/stenciltest/stenciltest.go (hunk 3 above)
Comment (jaredallard) [reply pkg/stenciltest/stenciltest.go:109]: Added `ErrorContains` for this usecase!

**Pull request merged**